Master v1 core

nmellado opened the pull request:
* **Please check if the PR fulfills these requirements**
- [x] Is the Pull-Request done against the right branch:
  - `master`: for hotfixes not impacting the API
  - `master-v1`: for changes related to the Radium Stable Release V1.
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

Be aware that the PR request cannot be accepted if it doesn't pass the Continuous Integration tests.


* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Update core structure according to `master-v1` recommendations

* **What is the current behavior?** (You can also link to an open issue here)
Fixes #302 


* **What is the new behavior (if this is a feature change)?**
See https://github.com/STORM-IRIT/Radium-Engine/issues/302#issuecomment-393505002


* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
Source files of Core are moved, and thus include directives need to be improved.

* **Other information**:

MathiasPaulin commented on src/Core/Animation/Handle/Skeleton.cpp:
```diff
@@ -239,18 +239,18 @@ std::ostream& operator<<( std::ostream& os, const Skeleton& skeleton ) {
                             ? "BRANCH"
                             : skeleton.m_graph.isLeaf( i ) ? "LEAF" : "UNKNOWN";
-        const int pid = skeleton.m_graph.m_parent.at( i );
+        const int pid = skeleton.m_graph.parents().at( i );
```
what is the objective of using `.at(i)` here and not `[i]`. Do we really need to do bound-checking at each access ?
The same remarks apply to several places below in `src/Core/Animation/Handle/Skeleton.cpp`

nmellado replied:
I do agree.
But I have not processed this file yet (only BVH and AdjacencyList + the side effects).

MathiasPaulin replied:
OK. This was just to ensure that you'll be attentive to this kind of stuff ;)

MathiasPaulin commented on src/GuiBase/BaseApplication.cpp:
```diff
@@ -122,9 +128,10 @@ BaseApplication::BaseApplication( int argc, char** argv, const WindowFactory& fa
+
+    std::stringstream ssTp;
+    ssTp << std::put_time(startTm, "%Y%m%d-%T");
+    QDir().mkdir( ssTp.str().c_str() );
```
As we are now using filesystem, perhaps it should be better to use ``std::filesystem::create_directory(ssTp.str()); `` instead of ``QDir... ``

nmellado replied:
You are right, I will.

MathiasPaulin commented on src/IO/AssimpLoader/AssimpFileLoader.cpp:
```diff
@@ -11,6 +11,15 @@
+#if __has_include(<filesystem>)
+include_filesystem
+#  include <filesystem>
+#  define my_filesystem std::filesystem
```
Do not #define aliases. Use ``using RaFilesystem=std::filesystem`` instead (or a name more simple if the aim of this alias is to simplify writing, e.g ``using stdfs=std::filesystem``) .
As this must be repeated in each file that need to acces ``std::filesystem``functionalities, put this include setting in one single file, e.g. ``Core/File/stdfilesystem.hpp``

nmellado commented on src/Core/Algorithm/Subdivision/CatmullClarkSubdivider.cpp:
```diff
@@ -138,7 +138,7 @@ bool CatmullClarkSubdivider::subdivide( TopologicalMesh& mesh, size_t n,
     // ###########################################################################
-    // FIXME: THIS IS ONLY THERE BECAUSE CORE ONLY MANAGES TRIANGLE MESHES FOR NOW
+    // FIXME: THIS IS ONLY THERE BECAUSE CORE ONLY MANAGES Vector3ui MESHES FOR NOW
 
```
thanks for spotting this, wrong copy/paste...
this is fixed now.

hoshiryu commented on src/Core/Mesh/MeshUtils.cpp:
```diff
@@ -118,7 +117,7 @@ RayCastResult castRay( const TriangleMesh& mesh, const Ray& ray ) {
 }
 
-/// Return the mean edge length of the given triangle mesh
+/// Return the mean edge length of the given Vector3ui mesh
```
Also restore triangle here too

hoshiryu commented on src/Core/Mesh/MeshUtils.hpp:
```diff
@@ -18,28 +18,28 @@ namespace MeshUtils {
 RA_CORE_API void getAutoNormals( TriangleMesh& mesh, VectorArray<Vector3>& normalsOut );
 
-/// Returns a list of edges from a given triangle mesh
+/// Returns a list of edges from a given Vector3ui mesh
```
here again

hoshiryu commented on src/Core/Mesh/MeshUtils.hpp:
```diff
@@ -51,10 +51,10 @@ struct RayCastResult {
 RA_CORE_API RayCastResult castRay( const TriangleMesh& mesh, const Ray& ray );
 
-/// Return the mean edge length of the given triangle mesh
+/// Return the mean edge length of the given Vector3ui mesh
```
here again

hoshiryu commented on src/Engine/Renderer/Mesh/Mesh.cpp:
```diff
@@ -127,7 +127,7 @@ void Mesh::loadGeometry( const Core::TriangleMesh& mesh ) {
 
 void Mesh::loadGeometry( const Core::Vector3Array& vertices, const std::vector<uint>& indices ) {
-    // Do not remove this function to force everyone to use triangle mesh.
+    // Do not remove this function to force everyone to use Vector3ui mesh.
```
here too

hoshiryu commented on src/Engine/Renderer/Mesh/Mesh.cpp:
```diff
@@ -139,9 +139,9 @@ void Mesh::loadGeometry( const Core::Vector3Array& vertices, const std::vector<u
     m_mesh.vertices() = vertices;
 
-    // Check that when loading a triangle mesh we actually have triangles or lines.
+    // Check that when loading a Vector3ui mesh we actually have triangles or lines.
```
here too

hoshiryu commented on src/Engine/Renderer/Mesh/Mesh.hpp:
```diff
@@ -102,7 +102,7 @@ class RA_ENGINE_API Mesh {
     MeshRenderMode getRenderMode() const { return m_renderMode; }
 
-    /// Returns the underlying triangle mesh.
+    /// Returns the underlying Vector3ui mesh.
```
here too

hoshiryu commented on src/Core/Geometry/Laplacian/Laplacian.hpp:
```diff
@@ -12,15 +11,15 @@ namespace Core {
 
 /**
- * For further reading on how to compute the Laplacian over a triangle mesh
+ * For further reading on how to compute the Laplacian over a Vector3ui mesh
```
here too

hoshiryu commented on src/Core/Geometry/Laplacian/Laplacian.hpp:
```diff
@@ -12,15 +11,15 @@ namespace Core {
  *
  *
- * Different approaches on data other than triangle meshes could be found in:
+ * Different approaches on data other than Vector3ui meshes could be found in:
```
here too

hoshiryu commented on src/Core/Mesh/MeshUtils.cpp:
```diff
@@ -158,20 +157,20 @@ void checkConsistency( const TriangleMesh& mesh ) {
     {
-        CORE_WARN_IF( !( getTriangleArea( mesh, t ) > 0.f ), "Triangle " << t << " is degenerate" );
-        const Triangle& tri = mesh.m_triangles[t];
+        CORE_WARN_IF( !( getTriangleArea( mesh, t ) > 0.f ), "Vector3ui " << t << " is degenerate" );
```
here too

hoshiryu commented on src/Core/Mesh/MeshUtils.cpp:
```diff
@@ -158,20 +157,20 @@ void checkConsistency( const TriangleMesh& mesh ) {
         {
             CORE_ASSERT( uint( tri[i] ) < mesh.vertices().size(),
-                         "Vertex " << tri[i] << " is in triangle " << t << " (#" << i
+                         "Vertex " << tri[i] << " is in Vector3ui " << t << " (#" << i
```
here too

hoshiryu commented on src/Core/Mesh/MeshUtils.cpp:
```diff
@@ -158,20 +157,20 @@ void checkConsistency( const TriangleMesh& mesh ) {
     for ( uint v = 0; v < visited.size(); ++v )
     {
-        CORE_ASSERT( visited[v], "Vertex " << v << " does not belong to any triangle" );
+        CORE_ASSERT( visited[v], "Vertex " << v << " does not belong to any Vector3ui" );
```
here too

hoshiryu commented on src/Core/Mesh/MeshUtils.hpp:
```diff
@@ -18,28 +18,28 @@ namespace MeshUtils {
 
-/// Fills the given array with the vertices of a given triangle.
-inline void getTriangleVertices( const TriangleMesh& mesh, TriangleIdx triIdx,
+/// Fills the given array with the vertices of a given Vector3ui.
```
here too

hoshiryu commented on src/Core/Mesh/MeshUtils.hpp:
```diff
@@ -18,28 +18,28 @@ namespace MeshUtils {
 
-/// Returns the area of a given triangle.
-inline Scalar getTriangleArea( const TriangleMesh& mesh, TriangleIdx triIdx );
+/// Returns the area of a given Vector3ui.
```
here too

hoshiryu commented on src/Core/Mesh/MeshUtils.hpp:
```diff
@@ -18,28 +18,28 @@ namespace MeshUtils {
 
-/// Computes the normal of a given triangle.
-inline Vector3 getTriangleNormal( const TriangleMesh& mesh, TriangleIdx triIdx );
+/// Computes the normal of a given Vector3ui.
```
here too

hoshiryu commented on src/Core/Mesh/MeshUtils.hpp:
```diff
@@ -18,28 +18,28 @@ namespace MeshUtils {
 
-/// Returns true if the triangle contains a given edge
-inline bool containsEdge( const Triangle& t1, uint v1, uint v2 );
+/// Returns true if the Vector3ui contains a given edge
```
here too

hoshiryu commented on src/Core/Mesh/MeshUtils.inl:
```diff
@@ -1,41 +1,45 @@
-    CORE_ASSERT( t1[0] == v2 || t1[1] == v2 || t1[2] == v2, "Vertex 2 not in triangle" );
+inline uint getLastVertex( const Vector3ui& t1, uint v1, uint v2 ) {
+    CORE_ASSERT( t1[0] == v1 || t1[1] == v1 || t1[2] == v1, "Vertex 1 not in Vector3ui" );
+    CORE_ASSERT( t1[0] == v2 || t1[1] == v2 || t1[2] == v2, "Vertex 2 not in Vector3ui" );
```
here too (twice)

hoshiryu commented on src/Core/Utils/StringUtils.hpp:
```diff
@@ -1,57 +1,36 @@
+/// @return the path to a parent directory of a given file's full path.
+/// Most of the time this function behaves similarly to the unix dirname(1)
+/// command. It will ignore any trailing slash.
+RA_CORE_API std::string getDirName( const std::string& path );
```
prefer multi-line comment

hoshiryu commented on src/Core/Utils/StringUtils.hpp:
```diff
@@ -1,57 +1,36 @@
+/// @param keepExtension If true it will behave like unix basename(1)
+/// (path/to/file.ext will give file.ext) ;
+/// otherwise it will remove the extension (path/to/file.ext will give file)
+RA_CORE_API std::string getBaseName( const std::string& path, bool keepExtension = true );
```
prefer multi-line comment

hoshiryu commented on src/Core/Utils/Timer.hpp:
```diff
@@ -1,31 +1,29 @@
+namespace Ra {
+namespace Core {
+// Convenience functions for timing.
+// Wrapping the highest resolution time measuring.
```
put in one line?

hoshiryu commented on src/Engine/Renderer/Mesh/Mesh.cpp:
```diff
@@ -139,9 +139,9 @@ void Mesh::loadGeometry( const Core::Vector3Array& vertices, const std::vector<u
+    // Check that when loading a Vector3ui mesh we actually have triangles or lines.
     CORE_ASSERT( m_renderMode != GL_TRIANGLES || nIdx % 3 == 0,
-                 "There should be 3 indices per triangle " );
+                 "There should be 3 indices per Vector3ui " );
```
here too

hoshiryu commented on src/Core/Math/LinearAlgebra.hpp:
```diff
@@ -158,6 +158,7 @@ template <typename Vector_>
 inline bool checkInvalidNumbers( Eigen::Ref<const Vector_> v, const bool FAIL_ON_ASSERT = false );
 
 /// Get two vectors orthogonal to a given vector.
+/// \warning fx must be normalized (this is not checked in the function)
```
prefer multi-line comment

hoshiryu commented on src/Core/Mesh/TriangleMesh.hpp:
```diff
@@ -84,29 +83,29 @@ class TriangleMesh {
     /// \note Attribute names "in_position" and "in_normal" are reserved and will
     ///       give an invalid handle.
```
prefer multi-line comment

hoshiryu commented on src/Core/Tasks/Task.hpp:
```diff
@@ -11,7 +11,7 @@ namespace Core {
 /// Tasks are basic processing units of the engine. Tasks are run
 /// in parallel on separate threads by the engine. The work of
 /// a task is defined by overriding its process() function.
```
prefer multi-line comment

hoshiryu commented on src/Core/Utils/Attribs.hpp:
```diff
@@ -174,43 +179,10 @@ class AttribManager {
 
     /// Copy all attributes from m.
     /// \note If some attrib already exists, it will be replaced.
```
prefer multi-line comment

hoshiryu commented on src/Core/Utils/StdUtils.hpp:
```diff
@@ -6,10 +6,10 @@
+namespace Utils {
 /// Helper function to combine multiple hashes by accumulation
 /// Called repeatedly to accumulate the hash value.
 /// Taken from boost.hash. http://www.boost.org/doc/libs/1_61_0/doc/html/hash.html
```
prefer multi-line comment

nmellado commented:
The PR is ready for review.

Note that I stopped trying to improve the quality of the code, and just followed the recommendation given at https://github.com/STORM-IRIT/Radium-Engine/issues/302#issuecomment-393505002. 
This PR is not aimed at fixing code styles, comments and doc, but simply to change the directory structure. These changes will have a strong impact on external codes (plugins, apps) and must be done before we release Radium v1. Other changes (performance improvements, code cleaning) can be done for the next release (minor versions).

Also, to have a better overview of the changes, I recommend to follow the changes one commit at a time. Each commit correspond to a simple change (moving files, changing namespace) + its effect on the other parts of the code.

Once this PR is accepted, we will have to work on the installation of redistribualble binaries, and we are done with v1 (except if I miss something ; please reply to this comment if you see something else).

hoshiryu commented on src/Core/Geometry/Adjacency/Adjacency.hpp:
```diff
@@ -45,24 +42,24 @@ RA_CORE_API AdjacencyMatrix uniformAdjacency( const uint point_size,
- * The function defined over the triangle is:
- *       f( i, j ) = 1 , if triangle i contains to vertex j
+ * The function defined over the Vector3ui is:
+ *       f( i, j ) = 1 , if Vector3ui i contains to vertex j
```
I still think this should be fixed in this PR since it has been introduced here.

hoshiryu replied:
sorry @nmellado , still there

hoshiryu commented on src/Core/Containers/BVH.hpp:
```diff
@@ -1,21 +1,32 @@
 template <typename T>
-class BVH {
+class [[deprecated( "Neither used nor tested" )]] BVH {
+    using Aabb = Eigen::AlignedBox<Scalar, 3>;
```
use it afterwards instead of `Eigen::AlignedBox<Scalar, 3>`

nmellado replied:
Fixed.

hoshiryu commented on src/CMakeLists.txt:
```diff
@@ -12,250 +12,9 @@ if ( "${isSysDir}" STREQUAL "-1" )
-if( RADIUM_ASSIMP_SUPPORT )
-    include_directories( ${ASSIMP_INCLUDE_DIR} )
-    message(STATUS "Compiling Radium with Assimp's include dir : ${ASSIMP_INCLUDE_DIR}" )
-endif( RADIUM_ASSIMP_SUPPORT )
```
these 2 messages about 3rdParty directories have been removed

nmellado replied:
Yes, this as been done in purpose.
The build chain will be updated in the next PR.

hoshiryu commented on src/Core/Animation/Animation.hpp:
```diff
@@ -1,22 +1,21 @@
 
-using KeyPose = std::pair<Scalar, Pose>;
-
 class RA_CORE_API Animation {
```
should it be merged with `Ra::Core::Animation::KeyPose` ?

nmellado replied:
As they were separated, I didn't want to merge them to avoid side effects.
It is true this needs to be fixed, but can be done later (should be).

nmellado commented on src/Engine/Renderer/RenderObject/RenderObject.hpp:
```diff
@@ -44,7 +44,7 @@ class RA_ENGINE_API RenderObject final : public Core::IndexedObject {
     RenderObject( const std::string& name, Component* comp, const RenderObjectType& type,
                   int lifetime = -1 );
-    ~RenderObject() override;
+    virtual ~RenderObject();
```
Some testing leftover, my bad.
Fixed.

nmellado commented on Plugins/MeshPaint/src/MeshPaintPlugin.cpp:
```diff
@@ -25,7 +25,7 @@ MeshPaintPluginC::MeshPaintPluginC() :
     m_PickingManager( nullptr ),
     m_system( nullptr ),
-    m_paintColor( Ra::Core::Color( 1.0, 0.0, 0.0, 1.0 ) ),
+    m_paintColor( Ra::Core::Utils::Color( 1.0, 0.0, 0.0, 1.0 ) ),
```
Done.

nmellado commented on src/Engine/Renderer/Renderers/ForwardRenderer.cpp:
```diff
@@ -360,7 +356,7 @@ void ForwardRenderer::debugInternal( const ViewingParameters& renderData ) {
                 shader->bind();
                 // lighting for Xray : fixed
-                shader->setUniform( "light.color", Ra::Core::Color(5.0, 5.0, 5.0, 1.0) );
+                shader->setUniform( "light.color", Ra::Core::Utils::Color( 5.0, 5.0, 5.0, 1.0 ) );
```
Done

nmellado commented on src/Engine/Renderer/Renderers/ForwardRenderer.cpp:
```diff
@@ -146,10 +145,10 @@ void ForwardRenderer::renderInternal( const ViewingParameters& renderData ) {
-    const auto clearZeros = Core::Colors::Black<Core::Colorf>();
-    const auto clearOnes = Core::Colors::FromChars<Core::Colorf>( 255, 255, 255, 255 );
-    const float clearDepth {1.0f};
+    const auto clearColor = Core::Utils::Color::Grey( Scalar( 0.0392 ) );
```
True.
Fixed by adding an optional parameter alpha to `Color::Grey`, here set to `0` (default is `1`).

dlyr replied:
Well, to have good transparent background screenshot, you need to do a special pass (to avoid color blend on background)

nmellado replied:
True.
But I just reverted to the previous version, and I'm looking forward your fancy rendering tricks coming with the new version ;).

MathiasPaulin replied:
The branch _brightness-fixes_ include some fixes in normalization coefficient for Blinn-Phong, default light intensity as well as fixes for good management of transparent background and transparency rendering. These fixes are important to have good rendering of all the assets one can load and use with the main app (from animation testing asset Dana, hand, ... to more general assets st sponza and others) as well as asset loaded by external plugins and must be integrated in master-v1 before release. So,  as soon as this branch is merged, I'll rebase brightness-fixes and will make a PR for this.

nmellado commented on src/GuiBase/Viewer/Viewer.cpp:
```diff
@@ -445,12 +450,12 @@ void Gui::Viewer::showEvent( QShowEvent* ev ) {
             new Engine::DirectionalLight( Ra::Engine::SystemEntity::getInstance(), "headlight" );
         // Set the default light intensity so that it does not burn the objects
-        light->setColor( Ra::Core::Color( 0.5, 0.5, 0.5, 1.0 ) );
+        // light->setColor( Ra::Core::Utils::Color( 0.5, 0.5, 0.5, 1.0 ) );
```
Done

nmellado commented:
Thanks @hoshiryu for your feedback.
I've addressed your comments and force-pushed the changes to get a clean history.
If you guys are happy with this PR, please don't forget to hit the `approve` button in the review form.
We may also wait the CI tests are passed, and then merge this.

MathiasPaulin commented:
There are some issues on MacOs Mojave. I'll fix them before reviewing this PR.

hoshiryu commented:
Actually trying to compile my own plugins based on this, will tell you if ok

nmellado commented on src/Core/Utils/Log.hpp:
```diff
@@ -5,8 +5,11 @@
-#include <Core/String/StringUtils.hpp>
+namespace Ra {
+namespace Core {
+namespace Utils {
```
You just need to do an `using namespace` in your code (not in headers of course) to use this.

nmellado commented on src/Core/Utils/Color.hpp:
```diff
@@ -0,0 +1,210 @@
+               ( uint32_t( scaled[2] ) << 8 ) | ( uint32_t( scaled[3] ) << 0 );
+    }
+
+    inline uint32_t toARGB32( const ColorBase<_Scalar>& color ) {
```
Fixed

nmellado commented on src/Core/Utils/Color.hpp:
```diff
@@ -0,0 +1,210 @@
+        return c;
+    }
+
+    inline uint32_t toRGBA32( const ColorBase<_Scalar>& color ) {
```
Fixed

hoshiryu commented:
All seems ok for me, except there is still the bad documentation modification in `Core/Geometry/Adjacency.hpp`: `triangle` has been replaced with `Vector3ui` (due to some replace-all I guess)

Also to be noted (linked to this PR due to split of `Core/Math/LinearAlgebra.hpp` ), there are 2 issues with the includes in `Core/Containres/Tex.inl`:
````
#include "Tex.hpp"
````
sould be changed to:
````
#include <Core/Containers/Tex.hpp>

#include <Core/Math/LinearAlgebra.hpp>
#include <Core/Math/Types.hpp>
````

hoshiryu commented on src/GuiBase/Viewer/Viewer.cpp:
```diff
@@ -445,12 +450,12 @@ void Gui::Viewer::showEvent( QShowEvent* ev ) {
             new Engine::DirectionalLight( Ra::Engine::SystemEntity::getInstance(), "headlight" );
         // Set the default light intensity so that it does not burn the objects
-        light->setColor( Ra::Core::Color( 0.5, 0.5, 0.5, 1.0 ) );
+        light->setColor( Ra::Core::Utils::Color::Grey( Scalar( 0.5 ) ) );
```
this should now be left as default `White()`, according to commit `b73775065b14c5a1a465827e51aa28c15fc95541`

nmellado replied:
This will be fixed later for v2, or next PR by @MathiasPaulin
